crates/matrix-sdk-indexeddb/src/stream.rs

```diff
@@ -0,0 +1,624 @@
+        // when the cursor is created. Because `Stream::poll_next` is designed the other
+        // way —no item is fetched before `poll_next` is called—, the first item must be
+        // stored somewhere. Here is a good place.
+        first_item: Option<(JsValue, JsValue)>,
+    }
+}
+
+impl<'a> StreamByCursor<'a> {
+    /// Build a new `StreamByCursor`.
+    ///
+    /// It takes a reference to an [`IdbObjectStore`].
+    ///
+    /// The cursor is opened by this method.
+    pub async fn new(object_store: &'a IdbObjectStore<'a>) -> Result<Self, DomException> {
+        let cursor = object_store.open_cursor()?.await?;
+
+        Ok(Self::new_with_cursor(cursor))
+    }
+
+    // Helper for `Self` and for `StreamByRenewedCursor`.
+    fn new_with_cursor(cursor: Option<IdbCursorWithValue<'a, IdbObjectStore<'a>>>) -> Self {
+        let first_item = cursor.as_ref().and_then(|cursor| cursor.key().zip(Some(cursor.value())));
+
+        Self { cursor, continue_cursor_future: None, first_item }
+    }
+}
+
+impl<'a> Stream for StreamByCursor<'a> {
+    type Item = Result<(JsValue, JsValue), DomException>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        // Let's get the cursor.
+        let Some(cursor) = &this.cursor else {
+            // No cursor? It means no matching records is found.
+            // There is no possible value. Let's close the `Stream`.
+            return Poll::Ready(None);
+        };
+
+        // A first result is ready. Let's return it.
+        // This operation happens once.
+        if let Some(value) = this.first_item.take() {
+            return Poll::Ready(Some(Ok(value)));
+        }
+
+        // The first value has been consumed. Let's poll for the next one?
+        //
+        // `continue_cursor` has not been called yet. Let's call it.
+        if this.continue_cursor_future.is_none() {
+            *this.continue_cursor_future = Some(Box::pin(cursor.continue_cursor()?));
+        }
+
+        // Now that `continue_cursor` has been called, we can poll its future.
+        //
+        // It returns a `bool`: `true` if a next value exists, `false` otherwise.
+        if ready!(this
+            .continue_cursor_future
+            .as_mut()
+            // SAFETY: The case where `continue_cursor_future` is `None` has been handled
+            // hereinabove.
+            .unwrap()
+            .as_mut()
+            .poll(cx))?
+        {
+            // At this point, we have a value ready to be returned. The
+            // `continue_cursor_future` can be dropped.
+            *this.continue_cursor_future = None;
+
+            Poll::Ready(Some(Ok((
+                cursor.key().unwrap_or_else(|| JsValue::UNDEFINED),
+                cursor.value(),
+            ))))
+        } else {
+            // It doesn't have a new value. End of the cursor. End of the stream.
```

Comment:
Ah, no, I see that the [ready macro](https://doc.rust-lang.org/std/task/macro.ready.html) returns `Poll::Pending` for us automatically.

Please could you add a comment somewhere explaining that, because it fooled me...